fdbclient/include/fdbclient/MetaclusterManagement.actor.h

```diff
@@ -1153,46 +1157,43 @@ struct CreateTenantImpl {
 
 			if (groupEntry.present()) {
 				ASSERT(groupEntry.get().assignedCluster.present());
-				if (self->preferAssignedCluster) {
-					ASSERT(groupEntry.get().assignedCluster.get() == self->tenantEntry.assignedCluster.get());
+				if (self->preferAssignedCluster &&
+				    groupEntry.get().assignedCluster.get() != self->tenantEntry.assignedCluster.get()) {
+					throw invalid_tenant_configuration();
 				}
 				return std::make_pair(groupEntry.get().assignedCluster.get(), true);
 			}
 		}
 
-		// Get a set of the most full clusters that still have capacity
-		// If preferred cluster is specified, look for that one.
 		state std::vector<Future<Reference<IDatabase>>> dataClusterDbs;
+		state std::vector<ClusterName> dataClusterNames;
 		state std::vector<Future<ClusterName>> clusterAvailabilityChecks;
-		if (self->preferAssignedCluster) {
-			dataClusterDbs.push_back(getAndOpenDatabase(tr, self->tenantEntry.assignedCluster.get()));
-			wait(waitForAll(dataClusterDbs));
-			clusterAvailabilityChecks.push_back(
-			    checkClusterAvailability(dataClusterDbs[0].get(), self->tenantEntry.assignedCluster.get()));
-		} else {
-			state KeyBackedSet<Tuple>::RangeResultType availableClusters =
-			    wait(ManagementClusterMetadata::clusterCapacityIndex.getRange(
-			        tr,
-			        {},
-			        {},
-			        CLIENT_KNOBS->METACLUSTER_ASSIGNMENT_CLUSTERS_TO_CHECK,
-			        Snapshot::False,
-			        Reverse::True));
-
-			if (availableClusters.results.empty()) {
-				throw metacluster_no_capacity();
-			}
+		// Get a set of the most full clusters that still have capacity
+		// If preferred cluster is specified, look for that one.
+		state KeyBackedSet<Tuple>::RangeResultType availableClusters =
+		    wait(ManagementClusterMetadata::clusterCapacityIndex.getRange(
+		        tr, {}, {}, CLIENT_KNOBS->METACLUSTER_ASSIGNMENT_CLUSTERS_TO_CHECK, Snapshot::False, Reverse::True));
+		if (availableClusters.results.empty()) {
+			throw metacluster_no_capacity();
+		}
 
-			for (auto clusterTuple : availableClusters.results) {
-				dataClusterDbs.push_back(getAndOpenDatabase(tr, clusterTuple.getString(1)));
+		bool preferredClusterPresent = false;
+		for (auto clusterTuple : availableClusters.results) {
+			if (self->preferAssignedCluster && self->tenantEntry.assignedCluster.get() != clusterTuple.getString(1)) {
```

Comment:
We are only checking a limited number of available clusters here, so it's possible the preferred cluster is valid but does not come back in this request.

I think it's probably most efficient to read the data cluster entry directly for the preferred cluster and check its capacity from the result. Otherwise you may need to read and scan the entire index to check whether the preferred cluster is present.